pkg/database/connection.go

```diff
@@ -0,0 +1,249 @@
+	// kube.Dialer
+	LookupFunc func(context.Context, string) ([]string, error)
+	// TLSConfig describes the TLS configuration for this database connection.
+	// nil disabled TLS connections.
+	// CA certs should be specified via RootCAs
+	// client certs should be specified via Certificates
+	TLSConfig *tls.Config
+	// ConnectTimeout restricts the whole connection process.
+	ConnectTimeout time.Duration
+	// Namespace that we are connecting to
+	Namespace string
+	// Context for the process
+	Context context.Context
+	// K8s Client
+	Client client.Client
+
+	// We use a dialer if we are not running inside K8s
+	RunningInsideK8s bool
+}
+
+func (c dbConfig) getClientTLSConfig(clientCertificateSecretName string, rootCertificateSecretName string) (tlsConfig *tls.Config, err error) {
+	r := resource.NewKubeResource(c.Context, c.Client, c.Namespace, kube.DefaultPersister)
+
+	tlsSecret, err := resource.LoadTLSSecret(clientCertificateSecretName, r)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to load client tls secret")
+	}
+
+	keyPair, err := tls.X509KeyPair(tlsSecret.Key(), tlsSecret.PriveKey())
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create key pair")
+	}
+
+	root, err := resource.LoadTLSSecret(rootCertificateSecretName, r)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to load root tls secret")
+	}
+
+	pool := x509.NewCertPool()
+	if !pool.AppendCertsFromPEM(root.CA()) {
+		return nil, fmt.Errorf("count not load CACert")
+	}
+
+	// Construct a tls.config
+	return &tls.Config{
+		Certificates: []tls.Certificate{keyPair},
+		RootCAs:      pool,
+	}, nil
+}
+
+func (c dbConfig) sslMode() string {
+	if c.TLSConfig == nil {
+		return "disable"
+	}
+
+	// TODO this is not used
+	if c.TLSConfig.InsecureSkipVerify {
+		return "verify-ca"
+	}
+
+	return "verify-full"
+}
+
+// openDB connects to a CockroachDB instance and verifies the connection
+func (c dbConfig) openDB() (*sql.DB, error) {
+	if c.Port == 0 {
+		c.Port = CockroachDBSQLPort
+	}
+
+	if c.Database == "" {
+		c.Database = "system"
+	}
+
+	connOptions := url.Values{
+		"sslmode": {c.sslMode()},
+	}
+
+	pgURL := url.URL{
+		Scheme:   "postgresql",
+		Host:     fmt.Sprintf("%s:%d", c.Host, c.Port),
+		Path:     fmt.Sprintf("/%s", c.Database),
+		RawQuery: connOptions.Encode(),
+	}
+
+	// pgx.Config's must be generated by ParseConfig
+	// otherwise connect will panic.
+	if c.User != "" {
+		pgURL.User = url.User(c.User)
+	}
+
+	pgCfg, err := pgx.ParseConfig(pgURL.String())
+	if err != nil {
+		return nil, err
+	}
+
+	if !c.RunningInsideK8s {
+		if c.DialFunc == nil {
+			return nil, errors.New("dial func is not set")
+		}
+		pgCfg.DialFunc = *c.DialFunc
+		pgCfg.LookupFunc = c.LookupFunc
+	}
+
+	pgCfg.TLSConfig = c.TLSConfig
+	pgCfg.ConnectTimeout = c.ConnectTimeout
+
+	db := stdlib.OpenDB(*pgCfg)
+
+	// Test the database connection
+	if _, err := db.Exec("SELECT 1"); err != nil {
+		return nil, err
```

Comment:
nit: I imagine that the errors from this will be incredible opaque. Might be worth tossing in an `errors.Wrap(err, "testing db connection")`

Reply:
Done